File: square_root/squareroot.py
def search(a, b, x):
    if abs(((a+b)/2)**2 - x) <= 0.001:
        return (a+b)/2
    elif ((a+b)/2)**2 < x:
        return search((a+b)/2, b, x)
    else:
        return search(a, (a+b)/2, x)
def root(x):
    if x == 0:
        return 0
    elif x == 1:
        return 1
    elif x < 0:
        return None
    else:
        return search(0, x, x)


print(root(9))

File: square_root/squareroot.hs
search :: Double -> Double -> Double -> Double
search a b x
    | abs (mid^2 -x) <= 0.001 = mid
    | mid^2 < x = search mid b x
    | otherwise = search a mid x
    where
        mid = (a + b) / 2

root :: Double -> Maybe Double
root x
    | x < 0 = Nothing
    | x == 0 = Just 0
    | x == 1 = Just 1
    | otherwise = Just (search 0 x x)


main :: IO()
main = print(root 12)